stellargraph/core/element_data.py

```diff
@@ -335,29 +381,62 @@ def __init__(self, ids, sources, targets, weights, type_starts):
         self._empty_ilocs = np.array([], dtype=np.uint8)
 
     def _init_directed_adj_lists(self):
-        # record the edge ilocs of incoming, outgoing and both-direction edges
-        in_dict = {}
-        out_dict = {}
-
-        for i, (src, tgt) in enumerate(zip(self.sources, self.targets)):
-            in_dict.setdefault(tgt, []).append(i)
-            out_dict.setdefault(src, []).append(i)
-
-        dtype = np.min_scalar_type(len(self.sources))
-        self._edges_in_dict = _numpyise(in_dict, dtype=dtype)
-        self._edges_out_dict = _numpyise(out_dict, dtype=dtype)
+        # record the edge ilocs of incoming and outgoing edges
+        if len(self.targets) == 0:
+            empty_array = np.array([], dtype=np.uint8)
+            self._edges_in_dict = FlatAdjacencyList(empty_array, empty_array)
+            self._edges_out_dict = FlatAdjacencyList(empty_array, empty_array)
+            return
+
+        def _to_dir_adj_list(arr):
+            neigh_counts = np.bincount(arr, minlength=self.number_of_nodes)
+            neigh_counts = neigh_counts.astype(self._id_index.dtype, copy=False)
+            splits = np.cumsum(neigh_counts, dtype=self._id_index.dtype)
```

Comment:
arg it looks like this still allocates a temp array! numpy is real sneaky with their inplace sounding functions 😡 

That being said this array is length `N` which is typically pretty negligable compared to `E`. Also I think any temp arrays will be cleaned up (maybe? hopefully?) by GC before the more expensive `np.argsort` stuff happens so this shouldn't affect peak memory.

I might've gotten it wrong but here's the code snippet i used to asses this:

```python
gc.collect()
neigh_counts = np.bincount(np.arange(10000))

tracemalloc.start()
mem_start = tracemalloc.take_snapshot()

splits = np.zeros(len(neigh_counts) + 1)
np.cumsum(neigh_counts, out=splits[1:])

gc.collect()
tracemalloc.stop()

mem = mem_snapshot_diff(mem_stop, mem_start)
```

Outputs 

```
print("Peak memory (MiB):", peak / (1024 * 1024))
print("Memory (MiB):", mem / (1024 * 1024))
print("ratio:", peak / mem ) 
```

Reply:
That doesn't record the peak memory or update the `mem_stop` snapshot? Also, it'd be great to include the actual output you see, so I can be sure I reproduce the numbers you see.

Simplifying the input data slightly and the moving the `splits` computation out, I see:

```python
import tracemalloc
import numpy as np

data = np.ones(1000000)
out = np.zeros(len(data) + 1)

tracemalloc.start()
np.cumsum(data, out=out[1:])
print(tracemalloc.get_traced_memory()) # (1606, 13846)

tracemalloc.stop()
```

That is, there doesn't seem to be any extra arrays created. The printed values don't change with larger arrays (e.g. adding an extra 0 or two).

Reply:
ah sorry i accidently copied the print statements instead of the output!

Reply:
In your example here, because you're only using `ones` bincount creates an intermediate array of size 2, and then copies this over. Using `np.arange` instead yields a ~1000x bigger peak memory usage!

```
import tracemalloc
import numpy as np

data = np.arange(1000000)
out = np.zeros(len(data) + 1)

tracemalloc.start()
np.cumsum(data, out=out[1:])
print(tracemalloc.get_traced_memory()) # (1606, 8_001_777)

tracemalloc.stop()
```

Reply:
Ah, good catch. Well, that's a bit dumb 😢

Let's file an issue against numpy?

Reply:
Real sneaky if you ask me! 😠 

Good idea, i'll file an issue.

Oh and I totally agree about the adding the extra 0, I just forgot to implement it! 😅

Reply:
I ended up filing https://github.com/numpy/numpy/issues/16281